### PR TITLE
Fixed #27039 -- Fixed fallback to model field default in model forms.

### PR DESCRIPTION
**Pull request by timgraham**
https://code.djangoproject.com/ticket/27039

### CODE REVIEW

File: django/forms/models.py [hunk 1]
```diff
@@ -52,6 +52,10 @@ def construct_instance(form, instance, fields=None, exclude=None):
             continue
         if exclude and f.name in exclude:
             continue
+        # Leave defaults for fields that aren't in POST data, except for
+        # checkbox inputs because they don't appear in POST data if not checked.
+        if f.has_default() and f.name not in form.data and not isinstance(form[f.name].field.widget, CheckboxInput):
```
Comment (timgraham) [inline django/forms/models.py]: I'm not too happy with this handling of checkboxes, but this at least starts the discussion...

File: docs/topics/forms/modelforms.txt [hunk 2]
```diff
@@ -332,6 +332,21 @@ Note that if the form :ref:`hasn't been validated
 ``form.errors``. A ``ValueError`` will be raised if the data in the form
 doesn't validate -- i.e., if ``form.errors`` evaluates to ``True``.
 
+If an optional fields doesn't appear in the form's data, the resulting model
```
Comment (cjerdonek) [inline docs/topics/forms/modelforms.txt]: optional field?

**Comment by timgraham** [comment]
@tomchristie, could you review?

**Comment by tomchristie** [comment]
Sure thing. Brief look all seems good, but will go through more thoroughly shortly.

File: django/forms/widgets.py [hunk 3]
```diff
@@ -481,6 +481,8 @@ def boolean_check(v):
 
 
 class CheckboxInput(Widget):
+    dont_use_model_field_default_for_empty_data = True
```
Comment (tomchristie) [inline django/forms/widgets.py]: Should we also include similar comment here?

**Comment by tomchristie** [comment]
Looks good.

**Comment by ryangallen** [comment]
@timgraham @claudep @AlexHill 

I'm experiencing a strange hiccup in the Django admin since upgrading to 1.10.1 that I think might be related to this issue. If it's not, I'm not sure what else could be causing it but it doesn't happen if I roll Django back to v1.9.9. Either way, I think you are the ones who can help demystify this for me. 

I have a field on a model set like so:
```verified = models.DateTimeField(null=True, blank=True, default=None)```

Any time a different value is entered for this field in the Django admin and saved, the field value on the object does not change. It remains the value it was set to before. If I remove the default value, the field saves normally. Any insight as to why this might be?

**Comment by timgraham** [comment]
See https://github.com/django/django/pull/7217.

**Comment by ryangallen** [comment]
Awesome, thanks much!

File: django/forms/models.py [hunk 4]
```diff
@@ -52,6 +52,11 @@ def construct_instance(form, instance, fields=None, exclude=None):
             continue
         if exclude and f.name in exclude:
             continue
+        # Leave defaults for fields that aren't in POST data, except for
```
Comment (RobertAKARobin) [inline django/forms/models.py]: @timgraham This caused me a bit of a debugging headache. I expected that if I did something like this:

```
def clean(self):
    self.cleaned_data['some_field'] = 'some_value'
```

...that would be sufficient to set what was ultimately saved to the database. But I found that `some_field` was always being saved with its default value. It took me a while to figure out why, from looking at the Django source. Now I'm doing this, with success:

```
def __init__(self, *args, **kwargs):
    super(MyForm, self).__init__(*args, **kwargs)
    self.data = self.data.copy()
    self.data['some_field'] = 'some_value'
```

I'm just curious what the reasoning is behind the rule you implemented here?

Comment (timgraham) [reply django/forms/models.py]: Your expectation seems reasonable. If there's a way to fix that while keeping backwards compatibility, feel free to offer a patch.

Comment (RobertAKARobin) [reply django/forms/models.py]: Do Django's automated tests ensure backward compatibility? If not I probably don't have enough familiarity with Django to do so. Glad I'm not taking crazy pills, though. :)

**Comment by timgraham** [comment]
If you have the tests passing that's a good start. Someone will review your patch as well.

**Comment by RobertAKARobin** [comment]
@timgraham Here we go: https://github.com/django/django/pull/11433

**Pull request merged**